src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -983,92 +985,145 @@ private static DocumentTombstone TableValueToTombstone(JsonOperationContext cont
             string key,
             long? expectedEtag,
             long? lastModifiedTicks = null,
-            ChangeVectorEntry[] changeVector = null)
+            ChangeVectorEntry[] changeVector = null,
+            LazyStringValue collection = null)
         {
-            // TODO: Should we call TableValueToDocument from here? Or it should be eascape the blittable building as of the data as we don't use it here?
-            var result = GetDocumentOrTombstone(context, loweredKey, throwOnConflict: false);
-            if (result.Item2 != null)
-                return null; //NOP, already deleted
+            var local = GetDocumentOrTombstone(context, loweredKey, throwOnConflict: false);
+            var currentLastTicks = DateTime.UtcNow.Ticks;
+            long etag = -1;
+            var collectionName = collection != null ? new CollectionName(collection) : null;
 
-            var doc = result.Item1;
-            long etag;
-            CollectionName collectionName;
-            if (doc == null)
+            if (_conflictCount > 0)
             {
-                if (_conflictCount > 0) //no tombstone and no document. perhaps we have a conflict?
+                if (local.Item2 != null && local.Item1 != null)
+                {
+                    // Something is wrong, we can't have conflicts and local document/tombstone
+                    throw new InvalidDataException($"we can't have conflicts and local document/tombstone with the key {loweredKey}");
+                }
+                var conflicts = GetConflictsFor(context, loweredKey);
+                if (conflicts.Count > 0) //we do have a conflict for our deletion candidate
+                {
+                    collectionName = ResolveConflictAndAddTombstone(context, changeVector, conflicts, out etag);
+                }
+            }
+            else
+            {
+
+                byte* lowerKey;
+                int lowerSize;
+                byte* keyPtr;
+                int keySize;
+
+                if (key != null)
                 {
-                    var conflicts = GetConflictsFor(context, loweredKey);
-                    if (conflicts.Count > 0) //we do have a conflict for our deletion candidate
+                    DocumentKeyWorker.GetLowerKeySliceAndStorageKey(context, key, 
+                        out lowerKey, out lowerSize, out keyPtr, out keySize);
+                
+                    if (local.Item2 == null && local.Item1 == null)
                     {
-                        collectionName = ResolveConflictAndAddTombstone(context, loweredKey, changeVector, conflicts, out etag);
-                        return new DeleteOperationResult
+                        // we adding a tombstone without having any pervious document, it could happened if this was called
+                        // from the incoming replication.
+
+                        if (expectedEtag != null)
+                            throw new ConcurrencyException(
```

Comment:
if I remove it WillVerifyEtags_OnDeleteNotThere test fail.
Since it tries to delete document that does not exists.